Updates to SelectBox

jaredlewis opened the pull request:
This adds a lot of updates to the select box component. @wesokes @jmcriffey can you teach me how to publish this on npm and if there is anything else i need to do?
- Make the search bar render inline with the component
- Automatically reposition the drop down to render above the select box if enough room is not available below
- Take into account when a value change occurs from a new prop being passed in
- If showClear is set to true replace the trigger with the clear button if a value exists

jaredlewis committed:
* Updates to selectbox

jaredlewis committed:
* Updates to select box

jaredlewis committed:
* Remove strict lints

jaredlewis committed:
* Lint

jaredlewis committed:
* Fix when updated value is null, reset the scrollTop for options on drop down hide

jeffrifwald commented on .eslintrc:
```diff
@@ -112,7 +112,6 @@
     "no-warning-comments": [1, {"terms": ["todo", "fixme"], "location": "anywhere"}],
     "no-with": 2,
     "padded-blocks": [2, "never"],
-    "prefer-const": 2,
```
Why did you take this out? This is deliberate.

jeffrifwald commented on .eslintrc:
```diff
@@ -135,9 +134,7 @@
     "space-before-blocks": [2, "always"],
     "space-in-parens": [2, "never"],
     "space-return-throw-case": 2,
-    "spaced-comment": [2, "never"],
```
Same here, explanation?

jaredlewis replied:
We always add spaces after comments so it seemed weird to do it differently

jeffrifwald replied:
You might, but this project does not. 

Sent from my iPhone

> On Mar 18, 2016, at 12:05 PM, Jared Lewis notifications@github.com wrote:
> 
> In .eslintrc:
> 
> > @@ -135,9 +134,7 @@
> >      "space-before-blocks": [2, "always"],
> >      "space-in-parens": [2, "never"],
> >      "space-return-throw-case": 2,
> > -    "spaced-comment": [2, "never"],
> >   We always add spaces after comments so it seemed weird to do it differently
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub

jeffrifwald commented on .eslintrc:
```diff
@@ -135,9 +134,7 @@
     "space-return-throw-case": 2,
-    "spaced-comment": [2, "never"],
     "use-isnan": 2,
-    "valid-jsdoc": 2,
```
Do you want invalid jsdocs?

jaredlewis replied:
Just trying to get this out the door and dont really have time to write full valid documentation

jeffrifwald replied:
That isn't how open source software works.

jaredlewis replied:
lol dude their were no comments to begin with so having some is better than none

jeffrifwald commented on src/SelectBox/components/SelectBox.js:
```diff
@@ -1,7 +1,7 @@
+import ReactDOM from 'react-dom';
 
 import {
-    BLUR_DELAY_MS,
```
How are you handling the blur delay?

jaredlewis replied:
It doesnt use blur to hide anymore, i added a document click

jeffrifwald replied:
Have you tested this method in every browser? I specifically did what I did to address small issues in several browsers. 

Sent from my iPhone

> On Mar 18, 2016, at 12:06 PM, Jared Lewis notifications@github.com wrote:
> 
> In src/SelectBox/components/SelectBox.js:
> 
> >  import {
> > -    BLUR_DELAY_MS,
> >   It doesnt use blur to hide anymore, i added a document click
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub

jeffrifwald commented:
Why did you change this component so much? Could you explain more? Why did you spend so much time updating things here without first starting a conversation?

-Make the search bar render inline with the component

What does this mean? I can't picture what you mean here.

-Automatically reposition the drop down to render above the select box if enough room is not available below

How are you determining if something has enough room? Sounds like a cool addition.

-Take into account when a value change occurs from a new prop being passed in

What does this mean? Shouldn't you just be passing in a new key to recreate the component instead of trying to reconcile 3 tiers of data (props, state, and input value)?

-If showClear is set to true replace the trigger with the clear button if a value exists

Why is this desirable? You want to force the user to clear before they are allowed to pick another value? I've never seen that before.

jaredlewis commented:
The search box renders next to the trigger instead of in the options list. It uses the positioning of the elements and the window size to determine where to position the drop down. You shouldnt have to redraw the whole component just to update the value, their may be some other internal state that you would lose, so i added in the ability to set it via prop. The clear button replaces the trigger but you can still click anywhere on the selector to show the drop down. All these were things we needed so i added them
